### PR TITLE
Improve the editor template workflow for users and devs

### LINKED ISSUE #1899: Remember preferred script template editor-wide instead of per-project
### Describe the project you are working on
Games in Godot

### Describe the problem or limitation you are having in your project
Surprisingly often it happens that I work on a new Godot project. Every single time whenever I create a new script, I forget to change the script template to empty and I end up with the cluttered default. This happens frequent enough that I'm writing this proposal right now .-.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
So basically the default script template should be saved for the editor, not per-project. Do we even support project-specific templates? With this, I can just set the template once and forget, like, forever. No matter how many new project I will be opening.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
See above.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Not without modifying the source.

### Is there a reason why this should be core and not an add-on in the asset library?
It's usability thing.

### LINKED ISSUE #3365: Add and suggest script templates based on the node/object type
### Describe the project you are working on

Physic improvement.

### Describe the problem or limitation you are having in your project

I wondering how to avoid common mistakes I often see in issues, plus, as `move_and_slide` has changed a lot in 4.0, limit the transition frustration.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Currently the template (which should perhaps be simplified)  is always the same:

Screenshot 2021-09-29 at 11 43 18

I think it will be great to add template according to the node.

For example, with the `CharacterBody`, we can add the minimal code to move a body:

   - get the direction
   - apply gravity
   - apply speed/acceleration
   - call move_and_slide

This is the most common code with this type of node.

Benefits:

- This will avoid common mistakes (like forgetting to store the result of move and slide) (and so issues)
- This will act as a mini tutorial for beginners
- This will save time for quick prototyping
- This will make the transition easier for 4.0

I thought about CharacterBody but this can be used for other nodes as well.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I'm not familiar with the editor code, but we can store the template code in a folder (/editor/template/xxx.gd), and simply read the template according to the classe. 

Each department could add his own node template if needed.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

no.

### Is there a reason why this should be core and not an add-on in the asset library?

it's core.

### LINKED ISSUE #28799: Add C# Script Templates for "Empty" and "No Comments"
GDScript already has three options for templates: "Default", "Empty", and "No Comments". This should also be added to C#.

Custom templates do already work when placed into `~/.config/godot/script_templates`, and I think that Godot simply needs to write the templates there as files. However, I took a look at the code for `CSharpLanguage::get_template()` and it appears to be hard-coded to only work with one template. I'm opening an issue rather than a PR to ensure it gets fixed properly.

Side note: Why are the GDScript templates defined in `editor/editor_settings.cpp` as opposed to something in `modules/gdscript`? Shouldn't the code be moved?

### LINKED ISSUE #36605: Script template should have a type requirement
**Godot version:**

ed0f1940cbdbe470da85f446f315ee12897038a1

**Issue description:**

When creating a new script and selecting a type, it doesn't always match the script template type.

**Steps to reproduce:**

Create a new script and set the type to Object.

### LINKED ISSUE #44534: Changing script language permanently changes default template
**Godot version:**

3.2.4 beta4

**Issue description:**

7oCVcTtEBU

### LINKED ISSUE #46436: EditorPlugins won't let you write "%BASE%"
**Godot version:** 3.2.4.rc2
**OS/device including version:** Windows 8.1
**Issue description:**
In my EditorPlugin I wrote the line ``file.replace("%BASE%",sceneRootNode.get_class())`` and whenever I restarted Godot I received the error ``Unable to load addon script from path: 'res://addons/plugin/plugin.gd' There seems to be an error in the code, please check the syntax.``
Attempting to re-enable the addon caused the same error. So I commented out the line but the error persisted.

**Steps to reproduce:**
Create a new plugin with the following lines and then restart Godot:
```
tool
extends EditorPlugin
# %BASE%
```

**Minimal reproduction project:**
[script_templates bug.zip](https://github.com/godotengine/godot/files/6047115/script_templates.bug.zip)

This bug isn't surprising to see (because of how script_templates work). I'm sure this one's a low priority issue, there are very easy workarounds such as writing "%"+"BASE"+"%". It doesn't seem to affect normal scripts but I haven't checked thoroughly.

### PR DESCRIPTION
**Pull request by fabriceci**
_Co-Authored-By: jmb462 <jmb462@gmail.com>_

This PR attempts to improve/rewrite the script template workflow to make the usage as easy/smooth as possible.

Closes https://github.com/godotengine/godot-proposals/issues/3365

fix #36605
fix #28799
fix #44534
fix #46436
close https://github.com/godotengine/godot-proposals/issues/1899

1_before-after

## What are the main changes?

- Built-in templates are in external file instead of hard-coded c++, they are no more mixed with user templates and can be updated
- Templates are now linked to a node, when you create a script for a node, you will only see the templates for this node, or one of its parents.
- Simplification and cleaning.

## How it can be used by core dev ##

The original motivation was to provide a more relevant template for CharacterBody nodes in order to :
    - help to avoid the common error we see in issues (like forget to assign the result of move and slide in 3.x)
    - help to make the transition in 4.0 (act like a tutorial)
    - fast prototyping, you have a moving character in few seconds

https://user-images.githubusercontent.com/6397893/137722735-84901d90-f322-4898-a986-0cb8129903c5.mp4

## How it works?

The template folders have the following structure

2_structure

To add a template simply create a file in the correct folder (`template_path/node_type/file.ext`), that's it.

By default:
 - the name of the template file will be the name of the file
 - the description will be empty
 - the space indent will be 4
 - the template will not be use as default
 
You can customize the template info with metas at the begning of your templates. Metas starts with the single-line comment syntax of your language:

```
# meta-name: Basic movement
# meta-description: Classic movement for gravity games (platformer, ...)
# meta-default: true
# meta-space-indent: 4
```

Example of description:

Screenshot 2021-10-18 at 14 40 43

**It works exactly the same for built-in and user templates**

## Notes

- When you create an editor plugin in c# you will have the basic structure for a plugin that I found in the doc. However there is a warning that pop up for now (I don't know why as I don't use c# & there was nothing before)
- I wonder if the **project** template folder should not be fixed like the add-ons one (not changeable in the editor), to avoid the ifdef in `gdscript.cpp`

### CODE REVIEW

**Review by aaronfranke — COMMENTED** [review]
I recommend reading the GDScript style guide: https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html

**Comment by pycbouh** [comment]
Not sure from the changes if you handle this or not, but I think templates at the root of the template folder (like now) can be handled as Object or Node templates for backward compatibility.

**Review by raulsntos — COMMENTED** [review]
So is the replacement syntax changed from `%BASE%` to `_BASE_`? Is there a reason to use `_` over `%`? It seems to me that `_` could be conflicting with valid scripts.

**Comment by raulsntos** [comment]
Good enough for me, I guess the chances of someone wanting to use a world like _CLASS_ in the output of the template are low.

**Comment by jmb462** [comment]
> They are suitable for all nodes though, just not all objects. Which is why I mentioned both. But since the syntax for placeholders was changed, I guess this is already a compatibility breaking change.

What I mean is if someone wrote a template suitable for all Control nodes (template include method calls from Control class) with the old system in the old root directory.

With the new system, we put it in a Control subfolder and it won't be proposed for Node2D or Node3D derivated nodes.

If this template is left in the root template directory and considered to be suitable for all Object, it will kill one of the avdandage of the new system (if a template is proposed, it is suitable).

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Amazing work!

I've left some comments about the physics templates and the general API in core, I'm not very familiar with specific script implementations.

**Review by raulsntos — CHANGES_REQUESTED** [review]

**Review by pouleyKetchoupp — APPROVED** [review]
Looks good to me, I've just commented on some minor things.

If possible, still would be good to have an extra approval for the csharp part.

File: modules/mono/editor_templates/CharacterBody2D/basic_movement.cs
Comment (raulsntos) [inline modules/mono/editor_templates/CharacterBody2D/basic_movement.cs:13]: This cast will throw an exception since it actually returns a `float` but the `GetSetting` returns an `object` so you can't cast from `object` to `int`, you have to cast to `float` first.

This would work:
```suggestion
    public int gravity = (int)(float)ProjectSettings.GetSetting("physics/2d/default_gravity");
```

File: modules/mono/editor_templates/CharacterBody2D/basic_movement.cs
Comment (raulsntos) [inline modules/mono/editor_templates/CharacterBody2D/basic_movement.cs:24]: The action `move_up` does not exist, in GDScript you are using `ui_accept`.

Same problem in the 3D template.

File: modules/mono/editor_templates/CharacterBody2D/basic_movement.cs
Comment (jmb462) [reply modules/mono/editor_templates/CharacterBody2D/basic_movement.cs]: Thanks, I've pushed the modification (but neither fabriceci nor me can currently test that C# templates are running well for a reason unrelated to this PR).

Comment (fabriceci) [reply modules/mono/editor_templates/CharacterBody2D/basic_movement.cs]: Nice spot, I tried with 3.X as 4.0 not works with mono for me, so I didn’t see the setting has changed from int to float!

**Comment by fabriceci** [comment]
@raulsntos Could you confirm now the mono templates work fine now ?

File: modules/mono/editor_templates/plugin.cs
Comment (raulsntos) [inline modules/mono/editor_templates/plugin.cs:6]: Missing the `partial` modifier here.

```suggestion
public partial class _CLASS_ : _BASE_
```

File: modules/mono/editor_templates/CharacterBody3D/basic_movement.cs
Comment (raulsntos) [inline modules/mono/editor_templates/CharacterBody3D/basic_movement.cs:29]: micro-nit: you may want to remove this space.

```suggestion
        Vector3 direction = Transform.basis.Xform(new Vector3(inputDir.x, 0, inputDir.y));
```

**Review by raulsntos — APPROVED** [review]
I can confirm the C# templates build fine now.

File: modules/gdscript/editor_templates/CharacterBody2D/basic_movement.gd
Comment (akien-mga) [inline modules/gdscript/editor_templates/CharacterBody2D/basic_movement.gd:23]: I understand why it's done that way, but I don't really like that we encourage a bad practice by reusing **UI** actions for character movement. Those should only be used for UI purposes, character movement should be done with new actions customized for each game, such as `move_left` or `steer_left`, etc.

Of course we don't want to add such gameplay-specific actions to the default InputMap, so I don't have a good alternative to propose.

File: modules/gdscript/gdscript_editor.cpp
Comment (akien-mga) [inline modules/gdscript/gdscript_editor.cpp:41]: Use relative paths:
```suggestion
#include "editor_templates/templates.gen.h"
```

File: modules/gdscript/gdscript_editor.cpp
Comment (akien-mga) [inline modules/gdscript/gdscript_editor.cpp:72]:
```suggestion
									 .replace(" -> void", "");
```
I don't think `→` is supported.

File: modules/mono/csharp_script.cpp
Comment (akien-mga) [inline modules/mono/csharp_script.cpp:59]:
```suggestion
#include "editor_templates/templates.gen.h"
```

**Review by akien-mga — CHANGES_REQUESTED** [review]

File: modules/gdscript/editor_templates/CharacterBody2D/basic_movement.gd
Comment (pouleyKetchoupp) [reply modules/gdscript/editor_templates/CharacterBody2D/basic_movement.gd:23]: What about adding a comment that explains what the good practice is (replace "ui_left" and "ui_right" with specific gameplay actions added to Project Settings > Input Mapping) so it still works out of the box but we encourage users to do things the proper way?

**Comment by mhilbrunner** [comment]
PR meeting: agree this looks good and we want to have this, maintainers/Akien need to figure out last details and then it's merge time :-) Good work!

File: editor/plugin_config_dialog.cpp
Comment (aaronfranke) [inline editor/plugin_config_dialog.cpp:295]: This is unrelated to this PR, but I noticed that this method should likely be changed to return `StringName`.

File: modules/mono/editor_templates/CharacterBody2D/basic_movement.cs
Comment (aaronfranke) [inline modules/mono/editor_templates/CharacterBody2D/basic_movement.cs:10]: This needs to be updated for consistency with the GDScript version.

File: modules/mono/editor_templates/CharacterBody3D/basic_movement.cs
Comment (aaronfranke) [inline modules/mono/editor_templates/CharacterBody3D/basic_movement.cs:10]: This needs to be updated for consistency with the changes I proposed (or whatever ends up being decided).

File: modules/mono/editor_templates/Node/default.cs
Comment (aaronfranke) [inline modules/mono/editor_templates/Node/default.cs:9]:
```suggestion
```

**Review by aaronfranke — CHANGES_REQUESTED** [review]
Overall looks quite good. Needs review from @vnen given the extent of the GDScript changes.

I noticed that `:=` doesn't get cleared with the other static types.

For the 3D character movement script, I'm tempted to suggest that there should be mouse movement for rotation, but rotating up/down is best done with a separate node for the head, so maybe it's out-of-scope for a one-script template. Or maybe we should just have `@onready var head: Node3D = $Head` and require users to provide a child node called Head (we already require a CollisionShape3D child for a character anyway). Or maybe we do both and have a separate script for this.

File: modules/mono/editor_templates/CharacterBody3D/basic_movement.cs
Comment (aaronfranke) [inline modules/mono/editor_templates/CharacterBody3D/basic_movement.cs:10]:
```suggestion
    public const float Speed = 5.0f;
    public const float JumpForce = 4.5f;
```

File: modules/mono/editor_templates/CharacterBody2D/basic_movement.cs
Comment (aaronfranke) [inline modules/mono/editor_templates/CharacterBody2D/basic_movement.cs:10]:
```suggestion
    public const float Speed = 300.0f;
    public const float JumpForce = -400.0f;
```

**Review by aaronfranke — CHANGES_REQUESTED** [review]
C# is picky about float types:

**Comment by fabriceci** [comment]
@aaronfranke @akien-mga Thanks for your reviews, all remarks should be addressed.
@aaronfranke For the `StringName` return for `get_name()`, if it's wanted, I can make another PR for that.

> For the 3D character movement script, I'm tempted to suggest that there should be mouse movement for rotation, but rotating up/down is best done with a separate node for the head, so maybe it's out-of-scope for a one-script template. Or maybe we should just have @onready var head: Node3D = $Head and require users to provide a child node called Head (we already require a CollisionShape3D child for a character anyway). Or maybe we do both and have a separate script for this.

I thought about this during the development of this RP, I think it is better to keep a template without the need to setup nodes to keep the thing simple and avoid confusion.

**Review by aaronfranke — APPROVED** [review]
This looks good to me. I tested the GDScript scripts and looked over the code. The Mono module doesn't work on the current master, but the scripts look good to me.

**Comment by KoBeWi** [comment]
The list of templates doesn't update when you change the base class. It maybe makes sense on scene, but when you create a script from file system dock you can't use any specific template.

If this was resolved, you could also add a proper template for EditorPlugin (right now it uses some weird special case) and a template for EditorScript.

File: core/object/script_language.h
Comment (KoBeWi) [inline core/object/script_language.h:304]: If the method returns vector of templates, shouldn't it be called `get_built_in_templates()`?

File: modules/gdscript/gdscript.cpp
Comment (KoBeWi) [inline modules/gdscript/gdscript.cpp:824]:
```suggestion
// Loading a template, don't parse.
```

**Comment by jmb462** [comment]
> The list of templates doesn't update when you change the base class. It maybe makes sense on scene, but when you create a script from file system dock you can't use any specific template.

Fixed.

>EDIT: From what I see, the editor will remember the template per each class. Which means that if you prefer empty template rather than default, you need to change it for every node type you use. That's very inconvenient.

A checkbox has been added to disable templates. It will use the empty template when it's unchecked. The checkbox will remember its state in the editor scope and for all nodes, so no need to disable it for each type of node.
 
> Not sure how could that be improved though. [The preferred template could be remembered for editor](https://github.com/godotengine/godot-proposals/issues/1899), but then project-specific templates are a problem.

The behaviour was updated:
 - the latest project-specific templates is stored at project level 
 - the latest preferred global templates are stored for the editor.

**Comment by KoBeWi** [comment]
> A checkbox has been added to disable templates. It will use the empty template when it's unchecked. The checkbox will remember its state in the editor scope and for all nodes, so no need to disable it for each type of node.

This seems to be saved only when Editor Settings are saved. I think when you confirm script creation and the checkbox state is different than it was before, the settings should automatically save.

Also, this makes the Empty template unnecessary, because you can just disable the checkbox.

> the latest preferred global templates are stored for the editor.

Doesn't seem to work.

btw what about EditorPlugin/EditorScript? Instead of hard-coded `plugin.gd`, you could use `EditorPlugin/default.gd`.

**Comment by fabriceci** [comment]
> btw what about EditorPlugin/EditorScript? Instead of hard-coded plugin.gd, you could use EditorPlugin/default.gd.

@KoBeWi Not sure I understand. Do you just suggest to move the current `plugin.gd/cs` to `EditorPlugin/default.gd` to match the other template structure?

**Comment by KoBeWi** [comment]
> Do you just suggest to move the current plugin.gd/cs to EditorPlugin/default.gd to match the other template structure?

Yes. This way we don't have a special case and the template can be used when creating new script from filesystem dock.

**Comment by jmb462** [comment]
> This seems to be saved only when Editor Settings are saved. I think when you confirm script creation and the checkbox state is different than it was before, the settings should automatically save. 
>[ global template saving] Doesn't seem to work.

Fixed. I didn't realise that settings were not saved immediatly. My tests were working because I was restarting the editor before swapping between projects. Now it's saved when the dialog is valided.

> Also, this makes the Empty template unnecessary, because you can just disable the checkbox.

I don't think so because someone could prefer empty template for a node type while using another templates for other nodes.
The checkbox will be used for those who don't care about template at all.

> btw what about EditorPlugin/EditorScript? Instead of hard-coded `plugin.gd`, you could use `EditorPlugin/default.gd`.

Done.

**Comment by fabriceci** [comment]
> Yes. This way we don't have a special case and the template can be used when creating new script from filesystem dock.

Thanks that was a good idea!

> Also, this makes the Empty template unnecessary, because you can just disable the checkbox.

It solves your issue: "if you prefer empty template rather than default, you need to change it for every node type you use. That's very inconvenient."

This setting applies the empty template to all nodes, on the other hand, in the template list, the choice is made by node.

@KoBeWi @akien-mga If there are no more comments, I think it's ready to be merged. All the requested changes are done.

**Comment by KoBeWi** [comment]
The dialog feels really cluttered now. I'd add separators here:
image
Also the Template dropdown could get disabled instead of disappearing when unchecking the "Use Template".

Also² you could add this template for EditorScript:
```GDScript
# meta-description: Basic editor script template
@tool
extends EditorScript

func _run():
    # Called when the script is executed (using File -> Run in Script Editor).
    pass
```

Other than that all looks fine now (still need to check the code).

**Comment by jmb462** [comment]
The dialog has been reorganized. The template part is now more compact.
The dropdown doesn't disapear any more.
EditorScript template has been added for GDScript and C#.

dialog

File: editor/script_create_dialog.cpp
Comment (KoBeWi) [inline editor/script_create_dialog.cpp:266]:
```suggestion
			// Save the last used template for this node into the project dictionary.
```
Misspelled dictionary. Same below.

**Comment by KoBeWi** [comment]
image
Disabled is redundant here; I see it's disabled.
It should say "Empty" or "No Template".

I get errors every time I open script create dialog:
image

File: editor/script_create_dialog.cpp
Comment (jmb462) [reply editor/script_create_dialog.cpp]: Fixed. Was a mix-up with the french spelling...

**Comment by jmb462** [comment]
> image Disabled is redundant here; I see it's disabled. It should say "Empty" or "No Template".
> 
> I get errors every time I open script create dialog.

Fixed. A test was missing before the first creation of the dictionary.

**Review by KoBeWi — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
